File: 백준/Silver/15654. N과 M （5）/N과 M （5）.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
nums = list(map(int, input().split()))
nums.sort()

v = [False] * (n + 1)

def backtrack(path):
  if len(path) == m:
    print(*path)
    return

  for i in range(len(nums)):
    if not v[i]:
      path.append(nums[i])
      v[i] = True
      backtrack(path)
      path.pop()
      v[i] = False

backtrack([])

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m;
	static int[] nums;
	static boolean[] v;
	static StringBuilder sb = new StringBuilder();
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		
		nums = new int[n];
		v = new boolean[n];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++) {
			nums[i] = Integer.parseInt(st.nextToken());			
		}
		Arrays.sort(nums);
		
		dfs(new ArrayList<>());
		System.out.println(sb);
	}
	
	static void dfs(List<Integer> path) {
		if(path.size() == m) {
			for(int i = 0; i < m; i++) {
				sb.append(path.get(i)).append(" ");
			}
			sb.append("\n");
			return;
		}
		
		for(int i = 0; i < n; i++) {
			if (!v[i]) {
				v[i] = true;
				path.add(nums[i]);
				dfs(path);
				v[i] = false;
				path.remove(path.size() - 1);
			}
		}
	}
}